### ISSUE TITLE
Implement better React key-assignment algorithm for controls

### ISSUE THREAD

**Issue by benwiley4000** [benwiley4000/react-responsive-audio-player#57]
Currently, we're just using array indices as keys for our rendered list of control components. But if the contents of our `controls` list changes on a component update, we should be able to determine which newly-listed controls match those which were previously mounted. This will be particularly important if a control instance is holding its own state, which we would prefer not to lose.

**Comment by benwiley4000** [comment]
Leaving on 2.0 milestone for now, but the same implementation should be applied in master after #47 is resolved, for the 1.2 release.

**Comment by benwiley4000** [comment]
Very much a work in progress... just trying to write these thoughts down somewhere:
```javascript
const newControls = [...nextProps.controls];
const oldControls = [...this.props.controls];
for (const control of newControls) {
  const matchingIndex = oldControls.indexOf(control);
  if (matchingIndex !== -1) {
    const matchingControl = oldControls.splice(matchingIndex, 1)[0];
    // this part is totally magic and we need a data structure for keeping
    // track of new keys/old keys (where multiple rendered instances could
    // share the same class). Obviously we can’t actually attach keys to
    // the class objects.
    control.key = matchingControl.key;
  } else {
    // we also need an algorithm for assigning new keys
    control.key = getNewKeyMagicallySomehow();
  }
}
```

**Comment by benwiley4000** [comment]
Boom!

```javascript
// somewhere
let nextControlKey = 0;
function getNextControlKey () {
  return (nextControlKey++).toString();
}

// in constructor
this.controlKeys = props.controls.map(getNextControlKey);

// in componentWillReceiveProps
const oldControls = [...this.props.controls];
this.controlKeys = nextProps.controls.map(control => {
  const matchingIndex = oldControls.indexOf(control);
  if (matchingIndex !== -1 && oldControls[matchingIndex]) {
    oldControls[matchingIndex] = null;
    return this.controlKeys[matchingIndex];
  }
  return getNextControlKey();
});
```

**Comment by benwiley4000** [comment]
Leaving open until included in master for 1.2 release.
